flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -3116,76 +3117,51 @@ mlir::Value IntrinsicLibrary::genSign(mlir::Type resultType,
+    return builder.createConvert(
+        loc, resultType,
+        fir::runtime::genSize(builder, loc, fir::getBase(array)));
 
-  // The value of the KIND argument is already reflected in the resultType
+  // Get the DIM argument.
+  mlir::Value dim = fir::getBase(args[1]);
 
-  mlir::Value result =
-      builder
-          .create<fir::BoxDimsOp>(loc, indexType, indexType, indexType, array,
-                                  zeroBasedDim)
-          .getResult(1);
-  return builder.createConvert(loc, resultType, result);
+  return builder.createConvert(
+      loc, resultType,
+      fir::runtime::genSizeDim(builder, loc, fir::getBase(array), dim));
 }
 
 // LBOUND
 fir::ExtendedValue
 IntrinsicLibrary::genLbound(mlir::Type resultType,
                             llvm::ArrayRef<fir::ExtendedValue> args) {
+  // Calls to LBOUND that don't have the DIM argument, or for which
+  // the DIM is a compile time constant, are folded to descriptor inquiries by
+  // semantics.  This function covers the situations where a call to the
+  // runtime is required.
   assert(args.size() == 3);
+  assert(!isAbsent(args[1]));
   if (const auto *boxValue = args[0].getBoxOf<fir::BoxValue>())
     if (boxValue->hasAssumedRank())
       TODO(loc, "LBOUND intrinsic with assumed rank argument");
 
-  // Calls to LBOUND that don't have the DIM argument, or for which
-  // the DIM is a compile time constant, are folded to descriptor inquiries by
-  // semantics.
-  assert(!isAbsent(args[1]));
   const fir::ExtendedValue &array = args[0];
-  llvm::SmallVector<mlir::Value> lbounds =
-      fir::factory::getNonDefaultLowerBounds(builder, loc, array);
-  if (lbounds.empty())
-    return builder.createIntegerConstant(loc, resultType, 1);
-  mlir::Type lbArrayType = fir::SequenceType::get(
-      {static_cast<fir::SequenceType::Extent>(array.rank())}, resultType);
-  auto lbArray = builder.createTemporary(loc, lbArrayType);
-  auto lbAddrType = builder.getRefType(resultType);
-  auto indexType = builder.getIndexType();
-  for (auto lb : llvm::enumerate(lbounds)) {
-    auto index = builder.createIntegerConstant(loc, indexType, lb.index());
-    auto lbAddr =
-        builder.create<fir::CoordinateOp>(loc, lbAddrType, lbArray, index);
-    mlir::Value lbValue = builder.createConvert(loc, resultType, lb.value());
-    builder.create<fir::StoreOp>(loc, lbValue, lbAddr);
-  }
-  mlir::Value dimArg = fir::getBase(args[1]);
-  mlir::Value one = builder.createIntegerConstant(loc, dimArg.getType(), 1);
-  mlir::Value zeroBasedDim =
-      builder.create<mlir::arith::SubIOp>(loc, dimArg, one);
-  mlir::Value resAddr =
-      builder.create<fir::CoordinateOp>(loc, lbAddrType, lbArray, zeroBasedDim);
-  return builder.create<fir::LoadOp>(loc, resAddr);
+  mlir::Value dim = fir::getBase(args[1]);
+  return builder.createConvert(
+      loc, resultType,
+      fir::runtime::genLboundDim(builder, loc, fir::getBase(array), dim));
```

Comment:
Lower bounds and descriptors are a bit treacherous. Outside of pointer and allocatable, one cannot assume  that the lower bounds of the entity are in the descriptor associated with the entity unless the descriptor was made locally.

With this change, the following program will print 1 instead of 3:
```
module m
contains
  subroutine foo(x, dim)
    integer :: x(2:, 3:, 4:), dim
    print *, lbound(x, dim)
  end subroutine
end module

use m
  integer :: x(7,8,9)
  call foo(x, 2)
end
```


The reason for that is that outside of pointer/allocatables, lower bounds are local to a scope, but the descriptor may come from outside the scope for dummy arguments, and we do not produce a new one in the current scope just to stick the lower bounds. Instead, non default local lower bounds are tracked in the ExtendedValue data structure and taken from there when needed (the call to `getNonDefaultLowerBounds` was retrieving them).

There is a way to ensure the local lower bounds are set in the descriptor that you are sending, which is to ensure the descriptor is made here for non-allocatable/pointers.

You would first need to change the `asBox` setting of `array` lowering to `asInquired` to avoid making a descriptor twice here.

Then you can do:
```
  mlir::Value box = array.match(
      [&](const fir::BoxValue &boxValue) -> mlir::Value {
        // This entity is mapped to a fir.box that may not contain the local
        // lower bound information if it is a dummy. Rebox it with the local
        // shape information.
        mlir::Value localShape = builder.createShape(loc, array);
        mlir::Value oldBox = boxValue.getAddr();
        return builder.create<fir::ReboxOp>(
            loc, oldBox.getType(), oldBox, localShape, /*slice=*/mlir::Value{});
      },
      [&](const auto &) -> mlir::Value {
        // This a pointer/allocatable, or an entity not yet tracked with a
        // fir.box. For pointer/allocatable, createBox will forward the
        // descriptor that contains the correct lower bound information. For
        // other entities, a new fir.box will be made with the local lower
        // bounds.
        return builder.createBox(loc, array);
      });
```
And to pass `box` to the runtime.

Reply:
Wow, Jean.  You're incredible.  Thanks for spotting this problem and giving me the solution.  You're the best.